text/0000-partial_types.md

```diff
@@ -0,0 +1,672 @@
+[guide-level-explanation]: #guide-level-explanation
+
+_**Note**: I didn't comment type explanations (as is needed for compiler) for colorizing purposes only._
+
+_**Note**: I use symbol `~` in code as a synonym of "equivalent" word only. It is not a Rust operator!_
+
+
+## Partial types by type Integrity
+
+```rust
+// case (A1)
+let foo : i16 = 0;
+```
+
+What is a **full type** of `foo`? Is it the `i16`? No, the `i16` is a sub-full_type in Rust type-system.
+```rust
+// case (A2)
+<variable>  : <full_type>;
+<variable>  : <type_clarification> <type>;
+<variable>  : <sharing> <'lifetime> <mutability> <type>;
+```
+
+If we explicitly write **full type** (using _unused_ Rust names) we get:
+```rust
+// case (A3)
+// FROM case (A1)
+foo  : value 'static const i16;
+```
+
+That mean, that variable `foo` has the `i16` (sub-)type with next type-clarification: `const`(not `mut`) mutability, `value`(not `&`) sharing and `'static` lifetime.
+
+I propose to extend type system by adding type integrity to sub-type. So, our variable will have next full type:
+```rust
+// case (A4)
+<variable>  : <type_clarification> <%integrity> <type>;
+<variable>  : <sharing> <'lifetime> <mutability> <%integrity> <type>;
+
+// case (A5)
+// FROM case (A1)
+foo  : value 'b const %a i16;
+foo  : value 'static const %full i16;
+foo  : %full i16;
+foo  : i16;
+```
+
+Lifetime variants are `'static` (for static lifetime), `'_`(don't care lifetime) and any other `'b`(some "b" lifetime).
+
+By the same analogy, integrity has similar names and meanings: `%full`(full integrity, soft keyword), `%_`(don't care how partial integrity is, soft keyword), `%empty` or `%!` (no integrity, soft keyword) and any other `%a`(some "a" integrity).
+
+Symbol `%` percent mean percent or part of the whole thing (variable in our case).
+
+_Note_: It is highly recommended to deprecate operator `%` as a remainder function (it is still no ambiguities to write "`\s+%\s+`"), and replace it with another operator (for example: `%mod` / `%rem` / `mod` / `rem`) to not to be confused by type integrity. 
+
+
+## Traits with Integrity variants
+
+We could already write Traits with **safe** virtual functions, that consumes virtual partial types having only variants of type integrity in Trait declaration
+```rust
+// case (B1)
+pub trait Upd {
+    type UpdType;
+
+    fn summarize<%a>(&self: & %a Self) -> String;
+
+    fn update_value<%a>(&mut self : &mut %a Self, newvalue: UpdType);
+
+    fn update_sametype<%a, %b>(&mut self : &mut %a Self, &another: & %b Self);
+}
+```
+
+Unfortunately, having variants of type integrity is not enough to write **safe** implementations or other non-virtual function declarations.
+
+
+## Detailed Integrity
+
+We need detailed integrity to write non-virtual specific typed parameters in function, including trait implementation.
+
+An abstractions is added for integrity detailing, we assume that **every** variable is a `struct`-like objects (even if it is not).
```

Comment:
What value do we gain from assuming that every variable is a struct? It doesn't really feel simpler to me, and all it requires is adding extra `.self` specifications to the end of every path.

Reply:
Good question.
1) We get unification
2) Enum allow unit-like values anyway (regardless if we want to represent numbers or not)